File: 056_MinimumWindowSubstring.cpp
// Topic: Sliding Window / Hashmap
// Minimum Window Substring: standard sliding window to cover all chars of t in s.
// Time Complexity: O(|S| + |T|)
// Space Complexity: O(1) assuming ascii / O(|T|)

#include <bits/stdc++.h>
using namespace std;

string minWindow(string s, string t)
{
    if (t.empty())
        return "";
    vector<int> need(128, 0);
    for (char c : t) need[c]++;
    int missing = t.size();
    int left = 0, minLen = INT_MAX, start = 0;
    for (int right = 0; right < s.size(); ++right)
    {
        if (need[s[right]] > 0)
            missing--;
        need[s[right]]--;
        while (missing == 0)
        {
            if (right - left + 1 < minLen)
            {
                minLen = right - left + 1;
                start = left;
            }
            need[s[left]]++;
            if (need[s[left]] > 0)
                missing++;
            left++;
        }
    }
    return minLen == INT_MAX ? "" : s.substr(start, minLen);
}

int main()
{
    cout << minWindow("ADOBECODEBANC", "ABC") << "\n";  // BANC
    cout << minWindow("a", "aa") << "\n";               // ""
    return 0;
}

File: 056_MinimumWindowSubstring.py
"""
Topic: Sliding Window / Hashmap
Problem: Minimum Window Substring
Description: Find smallest substring containing all chars of t in s.
Time: O(n)
Space: O(1) (alphabet) or O(|T|)
"""

from collections import Counter


def minWindow(s: str, t: str) -> str:
    if not t or not s:
        return ""
    need = Counter(t)
    missing = len(t)
    left = 0
    best = (float('inf'), None, None)
    for right, ch in enumerate(s):
        if need[ch] > 0:
            missing -= 1
        need[ch] -= 1
        while missing == 0:
            if right - left + 1 < best[0]:
                best = (right-left+1, left, right)
            need[s[left]] += 1
            if need[s[left]] > 0:
                missing += 1
            left += 1
    return "" if best[1] is None else s[best[1]:best[2]+1]


if __name__ == '__main__':
    print(minWindow("ADOBECODEBANC", "ABC"))  # BANC
    print(minWindow("a", "a"))  # a
